### PR TITLE
Development of 4 major functionalities

### PR DESCRIPTION
**Pull request by ShyamBhuptani**
Hi @flexplicateur  @kaushalnavneet 

This pull request is covering the functionalities of sensor fusion algorithm after step 6.
1. elimination of wrong data
2. weight coefficient
3. integrated support
4. parse for single time input
5. fused output

Please review and let me know if any changes required , so it can be implemented as earliest.

### CODE REVIEW

File: SensorFusion_src/elimination.c [hunk 1]
```diff
@@ -1,7 +1,41 @@
-double* elimination(double Z[], int m)
-{
-//returns a Z' containing only those sensor values whose value is supported by more than 70% sensors.
+/******************************************************
```
Comment (flexplicateur) [inline SensorFusion_src/elimination.c]: Will not compile in doxygen.

File: SensorFusion_src/main_test.c [hunk 2]
```diff
@@ -8,8 +8,19 @@
 #include <stdlib.h>
 #include "../include/support_degree_matrix.h"
 #include "../include/contribution.h"
+#include "../include/integrated_support.h"
+#include "../include/elimination.h"
+#include "../include/weight_coefficient.h"
+#include "../include/fused_output.h"
+#include "../include/parse.h"
+
 #define sdm_test_array_size 2
```
Comment (fcharih) [inline SensorFusion_src/main_test.c]: Constants are supposed to be ALL_CAPS.

File: SensorFusion_src/parse.c [hunk 3]
```diff
@@ -1 +1,76 @@
...
+*/
+
+/*Including dependancies*/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>	
+
+#include "../include/parse.h"	 
+//Number of sensors are fixed
+#define num_of_sensor 3
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Constants should be ALL_CAPS

File: SensorFusion_src/parse.c [hunk 4]
```diff
@@ -1 +1,76 @@
...
+
+/*Function declaration which returns a 2D array */
+double *Parse(char file_name[])
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Why does this function name start with a capital letter?

File: SensorFusion_src/parse.c [hunk 5]
```diff
@@ -1 +1,76 @@
...
+{
+	printf("file name is %s\n",file_name);
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Indentation????

File: SensorFusion_src/parse.c [hunk 6]
```diff
@@ -1 +1,76 @@
...
+#include <stdlib.h>
+#include <string.h>	
+
+#include "../include/parse.h"	 
+//Number of sensors are fixed
+#define num_of_sensor 3
+
+/*Function declaration which returns a 2D array */
+double *Parse(char file_name[])
+{
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Should not be on its own line, according to the coding convention.

File: SensorFusion_src/parse.c [hunk 7]
```diff
@@ -1 +1,76 @@
...
+	printf("file name is %s\n",file_name);
+    int iterations = 0;
+    double *sensor_readings = (double *) malloc(num_of_sensor * sizeof(double *));
+	printf("here1");
+    FILE* stream = fopen(file_name,"r");
+	printf("here2");
+	/*If stream is not present*/
+    if (!stream){
+        printf("Can't Open file");
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Why is "Open" capitalized? Where is the punctuation?

File: SensorFusion_src/parse.c [hunk 8]
```diff
@@ -1 +1,76 @@
...
+    }
+    char line[1024];
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: 1024 should be defined with a define macro.

File: SensorFusion_src/parse.c [hunk 9]
```diff
@@ -1 +1,76 @@
...
+	printf("here");
+    while(fgets(line,1024,stream)){
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Same, 1024 should not be hard-coded, but should be defined with a define macro.

File: SensorFusion_src/parse.c [hunk 10]
```diff
@@ -1 +1,76 @@
...
+        /*Using below condition removing the headers as the first row of
+		input file is assumed to be headers*/
+		printf("line is %s\n",line);
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Is this for debugging? If so, you should remove it.

File: SensorFusion_src/parse.c [hunk 11]
```diff
@@ -1 +1,76 @@
...
+		if(iterations == 0){
+            iterations++;
+            continue;
+        }
+        else{
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Bad indentation, should be:
```
if (foo) {
    bar();
} else {
    bacon();
}
```

File: SensorFusion_src/parse.c [hunk 12]
```diff
@@ -1 +1,76 @@
...
+            int count = 0;
+			/*Defining local variables*/
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Why are these comments weirdly indented?

File: SensorFusion_src/parse.c [hunk 13]
```diff
@@ -1 +1,76 @@
...
+            float time,sensor_value;
+            char * sensor_name;
+			/*delimiter will be comma as input is csv*/
+            char * delimiter = strtok(line,",");
+            while(delimiter){
+                if(count == 0){
+					/*First element of row is time
+					atof is used to convert time into a float*/
+                    time = (float)(atof(delimiter));
```
Comment (fcharih) [inline SensorFusion_src/parse.c]: Why do you cast? atof already returns a float.

**Review by fcharih — CHANGES_REQUESTED** [review]
Ok, so again, there are some important problems that you need to address:

1. Indentation needs a lot of work, especially the comments, which are all over the place.
2. Remove hard-coded values are use the `define` macro instead.
3. Your testing is not done correctly. You can't just print the function outputs to the console. You need to check if the values returned by the functions are the same as what you expect. The user should not have to visually compare the expected and actual outputs. 
4. Your documentation will not compile in Doxygen. You should go on the Doxygen website and read the instructions on how to write Doxygen-compliant documentation.
5. In a couple places, you do not follow the convention for curly braces and put them on their own line when starting a function. You should fix that.
6. You are not following the convention for header guards: the convention is `__FILENAME__`, go on CULearn for an example.

File: SensorFusion_src/main_test.c (hunk 2 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/main_test.c]: Done!!

File: SensorFusion_src/parse.c (hunk 4 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/parse.c]: Parse has been removed as it is fully created by amuleen.
this was just an interim development which worked for all inputs at a single time.

File: SensorFusion_src/parse.c (hunk 5 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/parse.c]: Parse has been removed as it is fully created by amuleen and will be added through develope_amuleen
this was just an interim development which worked for all inputs at a single time.

File: SensorFusion_src/parse.c (hunk 6 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/parse.c]: Parse has been removed as it is fully created by amuleen and will be added through develope_amuleen
this was just an interim development which worked for all inputs at a single time.

File: SensorFusion_src/parse.c (hunk 8 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/parse.c]: Got it!!

File: SensorFusion_src/parse.c (hunk 10 above)
Comment (ShyamBhuptani) [reply SensorFusion_src/parse.c]: Parse has been removed as it is fully created by amuleen and will be added through develope_amuleen
this was just an interim development which worked for all inputs at a single time.

**Review by ShyamBhuptani — COMMENTED** [review]
Parse has been removed as it is fully created by amuleen and will be added through develope_amuleen
this was just an interim development which worked for all inputs at a single time.
Added makefile and made some changes in test file
@flexplicateur  , your points 1,2,5 are resolved as parse.c was created for interim trial.
it has been removed now.
For others I tried to follow doxygen comments and header guard properly.

For testing, I guess it will not be a good idea just to print weather test is passed or fail as it can be hard-coded.
It will be the right approach if we print the output of function and expected output,
yes we can additionally compare the same which can be implemented in next commits.

**Comment by fcharih** [comment]
No, that's how testing is done. You hard-code the expected value and compare them with the output of the function that you're testing. The user should not have to visually compare all values. This is how unit testing works, and Dr. Ruiz mentioned it in class. Regardless, it's probably too late for you to change all the tests anyway.

I approve this PR. Good luck.

**Review by fcharih — APPROVED** [review]

**Pull request closed without merging**